File: Examples/Communication/CPP/Src/Server/Net/TcpServer.cpp
#include "Net/TcpServer.h"

namespace Communication
{
    namespace Net
    {
        class TcpServer::BindingInfo
        {
        private:
            boost::asio::io_service &IoService;
            std::shared_ptr<boost::thread> Task;
            Communication::BaseSystem::LockedVariable<bool> IsExited;

        public:
            std::shared_ptr<boost::asio::ip::tcp::acceptor> Acceptor;

            BindingInfo(boost::asio::io_service &IoService)
                : IoService(IoService),
                Task(nullptr),
                IsExited(false)
            {
            }

            ~BindingInfo()
            {
                IsExited.Update([](const bool &b) { return false; });
                if (Acceptor != nullptr)
                {
                    Acceptor->close();
                }
                if (Task != nullptr)
                {
                    Task->join();
                    Task = nullptr;
                }
            }

            void StartAccept(std::function<void(std::shared_ptr<boost::asio::ip::tcp::socket>)> AcceptHandler)
            {
                if (Task != nullptr) { throw std::logic_error("ThreadStarted"); }

                Task = std::make_shared<boost::thread>([=]()
                {
                    while (true)
                    {
                        if (IsExited.Check<bool>([](const bool &b) { return b; }))
                        {
                            return;
                        }
                        auto Socket = std::make_shared<boost::asio::ip::tcp::socket>(IoService);
                        boost::system::error_code se;
                        Acceptor->accept(*Socket, se);
                        if (se == boost::system::errc::success)
                        {
                            AcceptHandler(Socket);
                        }
                        else if (se == boost::system::errc::interrupted)
                        {
                            return;
                        }
                        else
                        {
                            auto Message = "UnexpectedError: " + se.message();
                            throw std::logic_error(Message);
                        }
                    }
                });
            }
        };

        TcpServer::TcpServer(boost::asio::io_service &IoService)
            : IoService(IoService),
            IsRunningValue(false),
            AcceptedSockets(std::make_shared<std::queue<std::shared_ptr<boost::asio::ip::tcp::socket>>>()),
            Sessions(std::make_shared<TSessionSet>()),
            IpSessions(std::make_shared<TIpAddressMap>()),
            StoppingSessions(std::make_shared<TSessionSet>()),
            BindingsValue(std::make_shared<std::vector<boost::asio::ip::tcp::endpoint>>()),
            SessionIdleTimeoutValue(Communication::BaseSystem::Optional<int>::CreateNotHasValue()),
            MaxConnectionsValue(Communication::BaseSystem::Optional<int>::CreateNotHasValue()),
            MaxConnectionsPerIPValue(Communication::BaseSystem::Optional<int>::CreateNotHasValue())
        {
        }

        TcpServer::~TcpServer()
        {
            Stop();
        }

        bool TcpServer::IsRunning()
        {
            return IsRunningValue.Check<bool>([](const bool &s) { return s; });
        }

        std::shared_ptr<std::vector<boost::asio::ip::tcp::endpoint>> TcpServer::GetBindings() const
        {
            return BindingsValue;
        }
        void TcpServer::SetBindings(std::shared_ptr<std::vector<boost::asio::ip::tcp::endpoint>> Bindings)
        {
            IsRunningValue.DoAction([&](bool &b)
            {
                if (b) { throw std::logic_error("InvalidOperationException"); }
                BindingsValue = Bindings;
            });
        }

        std::shared_ptr<Communication::BaseSystem::Optional<int>> TcpServer::GetSessionIdleTimeout() const
        {
            return SessionIdleTimeoutValue;
        }
        void TcpServer::SetSessionIdleTimeout(std::shared_ptr<Communication::BaseSystem::Optional<int>> ms)
        {
            IsRunningValue.DoAction([&](bool &b)
            {
                if (b) { throw std::logic_error("InvalidOperationException"); }
                SessionIdleTimeoutValue = ms;
            });
        }

        std::shared_ptr<Communication::BaseSystem::Optional<int>> TcpServer::GetMaxConnections() const
        {
            return MaxConnectionsValue;
        }
        void TcpServer::SetMaxConnections(std::shared_ptr<Communication::BaseSystem::Optional<int>> v)
        {
            IsRunningValue.DoAction([&](bool &b)
            {
                if (b) { throw std::logic_error("InvalidOperationException"); }
                MaxConnectionsValue = v;
            });
        }

        std::shared_ptr<Communication::BaseSystem::Optional<int>> TcpServer::GetMaxConnectionsPerIP() const
        {
            return MaxConnectionsPerIPValue;
        }
        void TcpServer::SetMaxConnectionsPerIP(std::shared_ptr<Communication::BaseSystem::Optional<int>> v)
        {
            IsRunningValue.DoAction([&](bool &b)
            {
                if (b) { throw std::logic_error("InvalidOperationException"); }
                MaxConnectionsPerIPValue = v;
            });
        }

        void TcpServer::DoAccepting()
        {
            while (true)
            {
                if (AcceptingTaskToken.IsCancellationRequested()) { return; }
                AcceptingTaskNotifier.WaitOne();
                while (true)
                {
                    std::shared_ptr<boost::asio::ip::tcp::socket> Socket = nullptr;
                    AcceptedSockets.DoAction([&](std::shared_ptr<std::queue<std::shared_ptr<boost::asio::ip::tcp::socket>>> &Sockets)
                    {
                        if (Sockets->size() > 0)
                        {
                            Socket = Sockets->front();
                            Sockets->pop();
                        }
                    });
                    if (Socket == nullptr)
                    {
                        break;
                    }

                    auto s = CreateSession();
                    s->NotifySessionQuit = [=]() { NotifySessionQuit(s); };
                    s->RemoteEndPoint = Socket->remote_endpoint();
                    s->IdleTimeout = SessionIdleTimeoutValue;
                    s->SetSocket(Socket);

                    if (MaxConnectionsValue->OnHasValue())
                    {
                        int SessionCount = Sessions.Check<int>([=](std::shared_ptr<TSessionSet> ss) -> int
                        {
                            return (int)(ss->size());
                        });

                        if (SessionCount >= MaxConnectionsValue->HasValue)
                        {
                            Communication::BaseSystem::AutoRelease Final([&]()
                            {
                                s->Stop();
                            });
                            s->Start();
                            if (MaxConnectionsExceeded != nullptr)
                            {
                                MaxConnectionsExceeded(s);
                            }
                            continue;
                        }
                    }

                    auto Address = s->RemoteEndPoint.address();
                    if (MaxConnectionsPerIPValue->OnHasValue())
                    {
                        int IpSessionCount = IpSessions.Check<int>([=](std::shared_ptr<TIpAddressMap> iss) -> int
                        {
                            return iss->count(Address) > 0 ? (*iss)[Address] : 0;
                        });

                        if (IpSessionCount >= MaxConnectionsPerIPValue->HasValue)
                        {
                            Communication::BaseSystem::AutoRelease Final([&]()
                            {
                                s->Stop();
                            });
                            s->Start();
                            if (MaxConnectionsPerIPExceeded != nullptr)
                            {
                                MaxConnectionsPerIPExceeded(s);
                            }
                            continue;
                        }
                    }

                    Sessions.DoAction([=](std::shared_ptr<TSessionSet> &ss)
                    {
                        ss->insert(s);
                    });
                    IpSessions.DoAction([=](std::shared_ptr<TIpAddressMap> &iss)
                    {
                        if (iss->count(Address) > 0)
                        {
                            (*iss)[Address] += 1;
                        }
                        else
                        {
                            (*iss)[Address] = 1;
                        }
                    });

                    s->Start();
                }
            }
        }

        void TcpServer::DoPurifiering()
        {
            while (true)
            {
                if (PurifieringTaskToken.IsCancellationRequested()) { return; }
                PurifieringTaskNotifier.WaitOne();
                while (true)
                {
                    std::shared_ptr<TcpSession> StoppingSession = nullptr;
                    StoppingSessions.DoAction([&](std::shared_ptr<TSessionSet> &Sessions)
                    {
                        if (Sessions->size() > 0)
                        {
                            StoppingSession = *(Sessions->begin());
                            Sessions->erase(StoppingSession);
                        }
                    });
                    if (StoppingSession == nullptr)
                    {
                        break;
                    }
                    auto Removed = false;
                    Sessions.DoAction([&](std::shared_ptr<TSessionSet> &ss)
                    {
                        if (ss->count(StoppingSession) > 0)
                        {
                            ss->erase(StoppingSession);
                            Removed = true;
                        }
                    });
                    if (Removed)
                    {
                        IpSessions.DoAction([&](std::shared_ptr<TIpAddressMap> &iss)
                        {
                            auto Address = StoppingSession->RemoteEndPoint.address();
                            if (iss->count(Address) > 0)
                            {
                                (*iss)[Address] -= 1;
                                if ((*iss)[Address] == 0)
                                {
                                    iss->erase(Address);
                                }
                            }
                        });
                    }
                    StoppingSession->Stop();
                }
            }
        }

        bool TcpServer::DoStopping(bool b)
        {
            if (!b) { return false; }

            for (int k = 0; k < (int)(BindingInfos.size()); k += 1)
            {
                auto bi = BindingInfos[k];
                bi->Acceptor->close();
            }
            BindingInfos.clear();

            if (AcceptingTask != nullptr)
            {
                AcceptingTaskToken.Cancel();
                AcceptingTaskNotifier.Set();
            }
            if (PurifieringTask != nullptr)
            {
                PurifieringTaskToken.Cancel();
                PurifieringTaskNotifier.Set();
            }
            if (AcceptingTask != nullptr)
            {
                AcceptingTask->join();
                AcceptingTask = nullptr;
            }
            if (PurifieringTask != nullptr)
            {
                PurifieringTask->join();
                PurifieringTask = nullptr;
            }

            Sessions.DoAction([=](std::shared_ptr<TSessionSet> &ss)
            {
                for (auto i = ss->begin(); i != ss->end(); i.operator ++())
                {
                    auto s = *i;
                    s->Stop();
                }
                ss->clear();
            });
            IpSessions.DoAction([=](std::shared_ptr<TIpAddressMap> &iss)
            {
                iss->clear();
            });

            return false;
        }

        void TcpServer::Stop()
        {
            IsRunningValue.Update([&](bool b) -> bool
            {
                return DoStopping(b);
            });
        }

        void TcpServer::Start()
        {
            bool Success = false;
            Communication::BaseSystem::AutoRelease ar([&]()
            {
                if (!Success)
                {
                    Stop();
                }
            });

            auto AcceptHandler = [this](std::shared_ptr<boost::asio::ip::tcp::socket> Socket)
            {
                this->AcceptedSockets.DoAction([=](std::shared_ptr<std::queue<std::shared_ptr<boost::asio::ip::tcp::socket>>> &Sockets)
                {
                    Sockets->push(Socket);
                });
                this->AcceptingTaskNotifier.Set();
            };

            auto AcceptingHandler = [=]() { DoAccepting(); };
            auto PurifieringHandler = [=]() { DoPurifiering(); };

            IsRunningValue.Update([&](bool b) -> bool
            {
                if (b) { throw std::logic_error("InvalidOperationException"); }

                if (BindingsValue->size() == 0)
                {
                    throw std::logic_error("NoValidBinding");
                }

                auto Exceptions = std::make_shared<std::vector<std::exception>>();
                for (int k = 0; k < (int)(BindingsValue->size()); k += 1)
                {
                    auto &Binding = (*BindingsValue)[k];
                    auto bi = std::make_shared<BindingInfo>(IoService);
                    try
                    {
                        auto Acceptor = std::make_shared<boost::asio::ip::tcp::acceptor>(IoService, Binding);
                        bi->Acceptor = Acceptor;
                        Acceptor->listen();
                    }
                    catch (std::exception &ex)
                    {
                        Exceptions->push_back(ex);
                        continue;
                    }
                    bi->StartAccept(AcceptHandler);
                    BindingInfos.push_back(bi);
                }
                if (BindingInfos.size() == 0)
                {
                    if (Exceptions->size() > 0)
                    {
                        throw std::logic_error((*Exceptions)[0].what());
                    }
                    else
                    {
                        throw std::logic_error("NoValidBinding");
                    }
                }

                AcceptingTask = std::make_shared<boost::thread>(AcceptingHandler);

                PurifieringTask = std::make_shared<boost::thread>(PurifieringHandler);

                Success = true;

                return true;
            });
        }
    }
}

File: Examples/Communication/CSharp/Src/Communication/Net/TcpServer.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using Communication.BaseSystem;

namespace Communication.Net
{
    public abstract class TcpServer<TServer, TSession> : IDisposable
        where TServer : TcpServer<TServer, TSession>
        where TSession : TcpSession<TServer, TSession>, new()
    {
        private class BindingInfo
        {
            public LockedVariable<Socket> Socket;
            public Task Task;
        }

        private LockedVariable<Boolean> IsRunningValue = new LockedVariable<Boolean>(false);
        public Boolean IsRunning
        {
            get
            {
                return IsRunningValue.Check(b => b);
            }
        }

        private Dictionary<IPEndPoint, BindingInfo> BindingInfos = new Dictionary<IPEndPoint, BindingInfo>();
        private CancellationTokenSource ListeningTaskTokenSource;
        private ConcurrentBag<Socket> AcceptedSockets = new ConcurrentBag<Socket>();
        private Task AcceptingTask;
        private CancellationTokenSource AcceptingTaskTokenSource;
        private AutoResetEvent AcceptingTaskNotifier;
        private Task PurifieringTask;
        private CancellationTokenSource PurifieringTaskTokenSource;
        private AutoResetEvent PurifieringTaskNotifier;
        private LockedVariable<HashSet<TSession>> Sessions = new LockedVariable<HashSet<TSession>>(new HashSet<TSession>());
        private LockedVariable<Dictionary<IPAddress, int>> IpSessions = new LockedVariable<Dictionary<IPAddress, int>>(new Dictionary<IPAddress, int>());
        private ConcurrentBag<TSession> StoppingSessions = new ConcurrentBag<TSession>();

        public TcpServer()
        {
        }

        private IPEndPoint[] BindingsValue = { };
        private int? SessionIdleTimeoutValue = null;
        private int? MaxConnectionsValue = null;
        private int? MaxConnectionsPerIPValue = null;

        public IPEndPoint[] Bindings
        {
            get
            {
                return BindingsValue;
            }
            set
            {
                IsRunningValue.DoAction
                (
                    b =>
                    {
                        if (b) { throw new InvalidOperationException(); }
                        BindingsValue = value;
                    }
                );
            }
        }
        public int? SessionIdleTimeout
        {
            get
            {
                return SessionIdleTimeoutValue;
            }
            set
            {
                IsRunningValue.DoAction
                (
                    b =>
                    {
                        if (b) { throw new InvalidOperationException(); }
                        SessionIdleTimeoutValue = value;
                    }
                );
            }
        }
        public int? MaxConnections
        {
            get
            {
                return MaxConnectionsValue;
            }
            set
            {
                IsRunningValue.DoAction
                (
                    b =>
                    {
                        if (b) { throw new InvalidOperationException(); }
                        MaxConnectionsValue = value;
                    }
                );
            }
        }
        public int? MaxConnectionsPerIP
        {
            get
            {
                return MaxConnectionsPerIPValue;
            }
            set
            {
                IsRunningValue.DoAction
                (
                    b =>
                    {
                        if (b) { throw new InvalidOperationException(); }
                        MaxConnectionsPerIPValue = value;
                    }
                );
            }
        }

        public void Start()
        {
            var Success = false;

            try
            {
                IsRunningValue.Update
                (
                    b =>
                    {
                        if (b) { throw new InvalidOperationException(); }

                        if (BindingsValue.Length == 0)
                        {
                            throw new Exception("NoValidBinding");
                        }

                        ListeningTaskTokenSource = new CancellationTokenSource();
                        AcceptingTaskTokenSource = new CancellationTokenSource();
                        AcceptingTaskNotifier = new AutoResetEvent(false);
                        PurifieringTaskTokenSource = new CancellationTokenSource();
                        PurifieringTaskNotifier = new AutoResetEvent(false);

                        var ListeningTaskToken = ListeningTaskTokenSource.Token;
                        var AcceptingTaskToken = AcceptingTaskTokenSource.Token;
                        var PurifieringTaskToken = PurifieringTaskTokenSource.Token;

                        var Exceptions = new List<Exception>();
                        foreach (var Binding in BindingsValue)
                        {
                            var Socket = new Socket(Binding.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

                            try
                            {
                                Socket.Bind(Binding);
                            }
                            catch (SocketException ex)
                            {
                                Exceptions.Add(ex);
                                continue;
                            }
                            Socket.Listen(MaxConnectionsValue.HasValue ? (MaxConnectionsValue.Value + 1) : 128);

                            var BindingInfo = new BindingInfo
                            {
                                Socket = new LockedVariable<Socket>(Socket),
                                Task = null
                            };
                            var Task = new Task
                            (
                                () =>
                                {
                                    try
                                    {
                                        while (true)
                                        {
                                            if (ListeningTaskToken.IsCancellationRequested) { return; }
                                            try
                                            {
                                                var a = BindingInfo.Socket.Check(s => s).Accept();
                                                AcceptedSockets.Add(a);
                                                AcceptingTaskNotifier.Set();
                                            }
                                            catch (SocketException)
                                            {
                                                if (ListeningTaskToken.IsCancellationRequested) { return; }
                                                BindingInfo.Socket.Update
                                                (
                                                    OriginalSocket =>
                                                    {
                                                        try
                                                        {
                                                            OriginalSocket.Close();
                                                        }
                                                        catch (Exception)
                                                        {
                                                        }
                                                        try
                                                        {
                                                            OriginalSocket.Dispose();
                                                        }
                                                        catch (Exception)
                                                        {
                                                        }
                                                        var NewSocket = new Socket(Binding.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
                                                        NewSocket.Bind(Binding);
                                                        NewSocket.Listen(MaxConnectionsValue.HasValue ? (MaxConnectionsValue.Value + 1) : 128);
                                                        Socket = NewSocket;
                                                        return NewSocket;
                                                    }
                                                );
                                            }
                                        }
                                    }
                                    catch (ObjectDisposedException)
                                    {
                                    }
                                },
                                TaskCreationOptions.LongRunning
                            );

                            BindingInfo.Task = Task;

                            BindingInfos.Add(Binding, BindingInfo);
                        }
                        if (BindingInfos.Count == 0)
                        {
                            throw new AggregateException(Exceptions);
                        }

                        AcceptingTask = new Task
                        (
                            () =>
                            {
                                while (true)
                                {
                                    if (AcceptingTaskToken.IsCancellationRequested) { return; }
                                    AcceptingTaskNotifier.WaitOne();
                                    while (true)
                                    {
                                        Socket a;
                                        if (!AcceptedSockets.TryTake(out a))
                                        {
                                            break;
                                        }

                                        var s = new TSession()
                                        {
                                            Server = (TServer)this,
                                            RemoteEndPoint = (IPEndPoint)(a.RemoteEndPoint)
                                        };
                                        if (SessionIdleTimeoutValue.HasValue)
                                        {
                                            a.ReceiveTimeout = SessionIdleTimeoutValue.Value * 1000;
                                        }
                                        s.SetSocket(a);

                                        if (MaxConnectionsValue.HasValue && (Sessions.Check(ss => ss.Count) >= MaxConnectionsValue.Value))
                                        {
                                            try
                                            {
                                                s.Start();
                                                if (MaxConnectionsExceeded != null)
                                                {
                                                    MaxConnectionsExceeded(s);
                                                }
                                            }
                                            finally
                                            {
                                                s.Stop();
                                                s.Dispose();
                                            }
                                            continue;
                                        }

                                        IPEndPoint e = (IPEndPoint)(a.RemoteEndPoint);
                                        if (MaxConnectionsPerIPValue.HasValue && (IpSessions.Check(iss => iss.ContainsKey(e.Address) ? iss[e.Address] : 0) >= MaxConnectionsPerIPValue.Value))
                                        {
                                            try
                                            {
                                                s.Start();
                                                if (MaxConnectionsPerIPExceeded != null)
                                                {
                                                    MaxConnectionsPerIPExceeded(s);
                                                }
                                            }
                                            finally
                                            {
                                                s.Stop();
                                                s.Dispose();
                                            }
                                            continue;
                                        }

                                        Sessions.DoAction
                                        (
                                            ss =>
                                            {
                                                ss.Add(s);
                                            }
                                        );
                                        IpSessions.DoAction
                                        (
                                            iss =>
                                            {
                                                if (iss.ContainsKey(e.Address))
                                                {
                                                    iss[e.Address] += 1;
                                                }
                                                else
                                                {
                                                    iss.Add(e.Address, 1);
                                                }
                                            }
                                        );

                                        s.Start();
                                    }
                                }
                            },
                            AcceptingTaskToken,
                            TaskCreationOptions.LongRunning
                        );

                        PurifieringTask = new Task
                        (
                            () =>
                            {
                                while (true)
                                {
                                    if (PurifieringTaskToken.IsCancellationRequested) { return; }

                                    PurifieringTaskNotifier.WaitOne();
                                    TSession StoppingSession;
                                    while (StoppingSessions.TryTake(out StoppingSession))
                                    {
                                        var Removed = false;
                                        Sessions.DoAction
                                        (
                                            ss =>
                                            {
                                                if (ss.Contains(StoppingSession))
                                                {
                                                    ss.Remove(StoppingSession);
                                                    Removed = true;
                                                }
                                            }
                                        );
                                        if (Removed)
                                        {
                                            var IpAddress = StoppingSession.RemoteEndPoint.Address;
                                            IpSessions.DoAction
                                            (
                                                iss =>
                                                {
                                                    if (iss.ContainsKey(IpAddress))
                                                    {
                                                        iss[IpAddress] -= 1;
                                                        if (iss[IpAddress] == 0)
                                                        {
                                                            iss.Remove(IpAddress);
                                                        }
                                                    }
                                                }
                                            );
                                        }
                                        StoppingSession.Stop();
                                        StoppingSession.Dispose();
                                    }
                                }
                            },
                            PurifieringTaskToken,
                            TaskCreationOptions.LongRunning
                        );

                        AcceptingTask.Start();
                        PurifieringTask.Start();

                        foreach (var BindingInfo in BindingInfos.Values)
                        {
                            BindingInfo.Task.Start();
                        }

                        Success = true;

                        return true;
                    }
                );
            }
            finally
            {
                if (!Success)
                {
                    Stop();
                }
            }
        }

        public void Stop()
        {
            IsRunningValue.Update
            (
                b =>
                {
                    if (!b) { return false; }

                    if (ListeningTaskTokenSource != null)
                    {
                        ListeningTaskTokenSource.Cancel();
                    }
                    foreach (var BindingInfo in BindingInfos.Values)
                    {
                        BindingInfo.Socket.DoAction
                        (
                            Socket =>
                            {
                                try
                                {
                                    Socket.Close();
                                }
                                catch (Exception)
                                {
                                }
                                try
                                {
                                    Socket.Dispose();
                                }
                                catch (Exception)
                                {
                                }
                            }
                        );
                    }
                    foreach (var BindingInfo in BindingInfos.Values)
                    {
                        if (BindingInfo.Task.Status != TaskStatus.Created)
                        {
                            BindingInfo.Task.Wait();
                        }
                        BindingInfo.Task.Dispose();
                    }
                    BindingInfos.Clear();
                    if (ListeningTaskTokenSource != null)
                    {
                        ListeningTaskTokenSource = null;
                    }

                    if (AcceptingTask != null)
                    {
                        AcceptingTaskTokenSource.Cancel();
                        AcceptingTaskNotifier.Set();
                    }
                    if (PurifieringTask != null)
                    {
                        PurifieringTaskTokenSource.Cancel();
                        PurifieringTaskNotifier.Set();
                    }
                    if (AcceptingTask != null)
                    {
                        AcceptingTask.Wait();
                        AcceptingTask.Dispose();
                        AcceptingTaskTokenSource.Dispose();
                        AcceptingTaskTokenSource = null;
                        AcceptingTask = null;
                    }
                    if (PurifieringTask != null)
                    {
                        PurifieringTask.Wait();
                        PurifieringTask.Dispose();
                        PurifieringTaskTokenSource.Dispose();
                        PurifieringTaskTokenSource = null;
                        PurifieringTask = null;
                    }

                    Sessions.DoAction
                    (
                        ss =>
                        {
                            foreach (var s in ss)
                            {
                                s.Stop();
                                s.Dispose();
                            }
                            ss.Clear();
                        }
                    );
                    IpSessions.DoAction
                    (
                        iss =>
                        {
                            iss.Clear();
                        }
                    );

                    if (AcceptingTaskNotifier != null)
                    {
                        AcceptingTaskNotifier.Dispose();
                        AcceptingTaskNotifier = null;
                    }
                    if (PurifieringTaskNotifier != null)
                    {
                        PurifieringTaskNotifier.Dispose();
                        PurifieringTaskNotifier = null;
                    }

                    return false;
                }
            );
        }

        public void NotifySessionQuit(TSession s)
        {
            StoppingSessions.Add(s);
            PurifieringTaskNotifier.Set();
        }

        public event Action<TSession> MaxConnectionsExceeded;
        public event Action<TSession> MaxConnectionsPerIPExceeded;

        public void Dispose()
        {
            Stop();
        }
    }
}
